iothub_client/tests/common_e2e/iothubclient_common_e2e.c

```diff
@@ -342,50 +368,101 @@ static void ReceiveUserContext_Destroy(EXPECTED_RECEIVE_DATA* data)
-    if (result != NULL)
+
+    if (result == NULL)
+    {
+        LogError("Failed allocating EXPECTED_SEND_DATA");
+    }
+    else
     {
         memset(result, 0, sizeof(*result));
+
         if ((result->lock = Lock_Init()) == NULL)
         {
             ASSERT_FAIL("unable to Lock_Init");
+            free(result);
+            result = NULL;
         }
         else
         {
-            char temp[1000];
-            char* tempString;
-            time_t t = time(NULL);
-            size_t string_length;
-            const char* data_fmt;
+            result->expectedString = string;
+            result->wasFound = false;
+            result->dataWasRecv = false;
+            result->result = IOTHUB_CLIENT_CONFIRMATION_ERROR;
+        }
+    }
 
-            if (g_e2e_test_options.use_special_chars)
-            {
-                data_fmt = TEST_EVENT_DATA_FMT_SPECIAL_CHAR;
-            }
-            else
-            {
-                data_fmt = TEST_EVENT_DATA_FMT;
-            }
+    return result;
+}
 
-            string_length = sprintf(temp, data_fmt, ctime(&t), g_iotHubTestId);
-            if ((tempString = (char*)malloc(string_length + 1)) == NULL)
-            {
-                Lock_Deinit(result->lock);
-                free(result);
-                result = NULL;
-            }
-            else
-            {
-                strcpy(tempString, temp);
-                result->expectedString = tempString;
-                result->wasFound = false;
-                result->dataWasRecv = false;
-                result->result = IOTHUB_CLIENT_CONFIRMATION_ERROR;
-            }
+static EXPECTED_SEND_DATA* EventData_Create(void)
+{
+    EXPECTED_SEND_DATA* result;
+    char temp[1000];
+    char* tempString;
+    time_t t = time(NULL);
+    size_t string_length;
+    const char* data_fmt;
+
+    if (g_e2e_test_options.use_special_chars)
+    {
+        data_fmt = TEST_EVENT_DATA_FMT_SPECIAL_CHAR;
+    }
+    else
+    {
+        data_fmt = TEST_EVENT_DATA_FMT;
+    }
+
+    string_length = sprintf(temp, data_fmt, ctime(&t), g_iotHubTestId++);
+    if ((tempString = (char*)malloc(string_length + 1)) == NULL)
+    {
+        result = NULL;
+    }
+    else
+    {
+        (void)strcpy(tempString, temp);
+        result = EventData_Create_With_String(tempString);
+
+        if (result == NULL)
+        {
+            free(tempString);
         }
     }
+
+    return result;
+}
+
+static EXPECTED_SEND_DATA* EventData_Create_With_Custom_Size(size_t messageSize)
+{
+    EXPECTED_SEND_DATA* result;
+    char* message = malloc(sizeof(char) * messageSize);
+
+    if (message == NULL)
+    {
+        LogError("Failed to allocate EventData message");
+        result = NULL;
+    }
+    else
+    {
+        char idString[10];
+        size_t idStringLength = sprintf(idString, "%d", (int)g_iotHubTestId++);
+
+        (void)memset(message, (int)'a', messageSize - 1);
```

Comment:
So if I'm reading this right, it looks like the message will be something like "a<whatever g_iotHubTestId set>aaaaaa... (up to like 64KB, depending on caller)".  Correct?

Having the memcpy of the `idstring` on the line immediately after the `memset` of all the a's while OK since it's overwriting memory we own gave me pause for a bit.  Does the 1st byte *have* to be an a?

So (I have NOT vetted this for all the offby one) IMO below is a bit easier to understand since we're not overwriting ourselves.

```c
memcpy(message, idString, idStrengLength);
memset(message+idStringLength, 'a', messageSize -idStringLength - 1);
```

This is more a nit/consider, just in general and maybe i'm missing something.

Reply:
Makes sense. Done.